src/Lean/Server/CodeActions.lean

```diff
@@ -0,0 +1,88 @@
+/-
+Copyright (c) 2022 E.W.Ayers. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+
+Authors: E.W.Ayers
+-/
+
+import Lean.Server.FileWorker.RequestHandling
+import Lean.Server.InfoUtils
+
+namespace Lean.Server
+
+open Lsp
+open RequestM
+open Snapshots
+
+/-- A code action provider is a function for getting LSP code actions for vscode. You can register them with the `@[codeActionProvider]` attribute.
+
+This is a low-level interface for making LSP code actions.
+This interface can be used to implement the following applications:
+- refactoring code: adding underscores to unused variables,
+- suggesting imports
+- document-level refactoring: removing unused imports, sorting imports, formatting.
+- Helper suggestions for working with type holes (`_`)
+- Helper suggestions for tactics.
+
+When implementing your own `CodeActionProvider`, we assume that no long-running computations are allowed.
+If you need to create a code-action with a long-running computation, we suggest the following approach:
```

Comment:
This feels a bit disappointing to be honest. Would it be too complicated to support `codeAction/resolve` from the beginning? Tbh I think this would have been the more reasonable LSP API from the beginning.

Reply:
Ah, I missed your edit regarding future work. Perhaps it would be enough to point that out in the docstring then. I just want to avoid the macro approach becoming dominant where an async edit would be the more user-friendly choice.

Reply:
The `codeAction/resolve` API looks a bit strange since it is a client->server request. I would expect a server->client notification or request that we could use to just push the result to the client once computed, but as specified it seems like the client has to poll for the result. Is that what's supposed to happen @EdAyers ? I am guessing there could be a `codeActionResults` table or something like that in the server which the code action providers store their results into once finished, and `codeAction/resolve` looks up values in that table.

Reply:
I can add `codeAction/resolve` to this PR. I wasn't sure how to support this though. I guess `CodeActionProvider`'s return type should support a lazy variant of `CodeAction`?

Reply:
@Vtec234 The whole idea of `codeAction/resolve` is that the code action is only computed if the user selects it.  It works like this:
1. LSP server sends list of possible code actions.
2. User selects a code action.
3. Editor requests the edit for the code action via `codeAction/resolve`.

Reply:
I mean, what should the Lean-side API be? We don't want code action writers to handle both `textDocument/codeAction` and `codeAction/resolve`, so there should be some system for matching up the `codeAction/resolve` requests with the original `textDocument/codeAction` requests.

Reply:
If taking the strategy of getting the server to store the array, I'm not sure how the data should be stored. I think the options are:
- `RequestM` is modified to have a state
- Store the state as an `IO.ref`

Reply:
I'm not concerned about the implementation, but about storing the *last* code action request specifically.  Does vscode never send a codeAction request while the popup menu is open?  Do other editors behave in the same way?

It's not particularly bad if we get this wrong now, this doesn't affect the editor API and users are supposed to use higher-level APIs instead.

Reply:
I'm making the assumption that `codeAction/resolve` will only be sent for the most recent result of `textDocument/codeAction`. [The spec](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#codeAction_resolve) doesn't say that this is necessary, but I don't think there are common cases where you would want to resolve something that wasn't from the most recent request.

Reply:
Code action requests are not only sent due to explicit user actions.  I believe VS Code queries the code actions for the whole file so that it knows where to show the lightbulbs (or maybe just the currently visible part of the file, I never looked too much into it).

So something like the following does not seem impossible:
1. User does ctrl+dot (does this send a separate code action request?)
2. VS Code updates the lightbulbs with a code action request.
3. After a few seconds of decision making, the user selects the second code action, returned by (1), which is no longer the last request.

Reply:
@gebner That's really good to know! Then I agree that `resolve` should be a separate function of a `CodeActionProvider`, with no state being stored between their invocations.